### DISCUSSION TITLE
virtual user 1 stuck in loading item

### DISCUSSION THREAD

**Discussion by parisstc** [TPC-Council/HammerDB#553] [Q&A] [answered]
Hello,
I am trying to find what I am doing wrong but I have run out of ideas, I have a remote server running mysql and would like to test its performance with hammerdb,
So I have downloaded the tar.gz on my linux host and installed the necessary libs with
sudo apt-get install libxft-dev
sudo apt-get install libcairo2-dev
When I start hammerdb and run librarycheck, mysql looks ok:
`HammerDB CLI v4.7
Copyright (C) 2003-2023 Steve Shaw
Type "help" for a list of commands
Initialized SQLite on-disk database /tmp/hammer.DB using existing tables (36,864 KB)
hammerdb>librarycheck
Checking database library for Oracle
Error: failed to load Oratcl - can't read "env(ORACLE_HOME)": no such variable
Ensure that Oracle client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MSSQLServer
Error: failed to load tdbc::odbc - couldn't load file "libiodbc.so": libiodbc.so: cannot open shared object file: No such file or directory
Ensure that MSSQLServer client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for Db2
Error: failed to load db2tcl - couldn't load file "/home/parisstc/HammerDB-4.7/lib/db2tcl2.0.1/libdb2tcl.so": libdb2.so.1: cannot open shared object file: No such file or directory
Ensure that Db2 client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MySQL
Success ... loaded library mysqltcl for MySQL
Checking database library for PostgreSQL
Error: failed to load Pgtcl - couldn't load file "/home/parisstc/HammerDB-4.7/lib/pgtcl2.1.1/libpgtcl2.1.1.so": libpq.so.5: cannot open shared object file: No such file or directory
Ensure that PostgreSQL client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MariaDB
Error: failed to load mariatcl - couldn't load file "/home/parisstc/HammerDB-4.7/lib/mariatcl0.1/libmariatcl0.1.so": libmariadb.so.3: cannot open shared object file: No such file or directory
Ensure that MariaDB client libraries are installed and the location in the LD_LIBRARY_PATH environment variable

hammerdb>
`
But when I run the tool, although the tpcc database with its tables is created, it gets stuck forever in the state shown in the attached screenshot and the stored procedures etc never get created:
![image](https://github.com/TPC-Council/HammerDB/assets/135003804/472371c5-77f0-4cc1-9d17-c5a8882691d6)

I couldnt find anything in the logfile, mysql processlist shows nothing useful,
could you please help me on how to troubleshoot this?

**Comment by parisstc** [comment]
I decided to let it run for as long as possible,
It took more than 1 hour for the "TPCC SCHEMA COMPLETE" message, is this normal?
the host is an openstack vm with 4 cores and 8 gig of ram

**Comment by sm-shaw** [comment, marked as answer]
So the key bottleneck here is the network round trip between the client and the server whilst doing the load, so yes if this network round trip is not particularly fast then the load can and will take a long time. 

By default, the HammerDB schema creation and load has been designed to be as easy as possible, so you start the schema build running, and it will create and load the data dynamically without intermediate files. It does this by calling standard 'insert' calls. For the larger tables mySQL uses multi-row inserts, for the item table because it is smaller and usually finishes much faster we insert a row at a time - so as you see the item load taking longer, then we know the issue is the network round trip as it is not using multi-row.  Additionally, MySQL networking is not as performant as other databases, so the impact is amplified with MySQL. 

Note that even loading locally, there is always the overhead of the database maintaining consistency by logging all of your transactions in case of recovery, so most of the load overhead is on the database side (for all databases). 

For these reasons, HammerDB supports a [datagen](https://www.hammerdb.com/docs/ch13.html) option - this allows you to create the data locally as flat files, this will run much faster than even a local database load as it is only dependent on the local CPU and disk performance. You can then ship these files to the remote server and do a bulk load that is not logged. The link above explains how to do this and complete the schema build for each database. This is more manual overhead initially, but much quicker to do the actual bulk load itself. So if you are doing repeated builds and/or working remotely this is the approach to take and the reason this datagen feature was added. 

Also be aware that if testing remotely, then the network round trip will impact the test performance as well. With MySQL for reasons noted above you will need a larger number of Virtual Users than testing locally to reach equivalent performance each on a LAN. The highest performance for MySQL will come through connecting through a local socket connection.

**Reply by parisstc** [reply]
thank you for the prompt reply
